Support []string and []*string as sensitive fields

### What problem are you facing?

Tring to generate a CRD with sensitive field type []*string fails with:

```
panic: cannot generate crd: cannot build types for User: cannot build the types: got type "[]*string" for field "Passwords", only types "string" and "*string" supported as sensitive
```

```
config.Store.SetForResource("aws_elasticache_user", config.Resource{
		ExternalName: config.ExternalName{
			DisableNameInitializer: true,
		},
		Sensitive: config.Sensitive{
			CustomFieldPaths: []string{
				"passwords", // interestingly this field is not marked as sensitive in tf schema
			},
		},
		UseAsync: true,
	})
```

### How could Terrajet help solve your problem?

Support []string and []*string as sensitive fields

In GCP, I got the following for `google_storage_bucket_object`:

```console
panic: cannot generate crd for resource google_storage_bucket_object: cannot build types for BucketObject: cannot build the Types: got type "[]github.com/crossplane-contrib/provider-tf-gcp/apis/storage/v1alpha1.CustomerEncryptionParameters" for field "CustomerEncryption", only types "string" and "*string" supported as sensitive
```

However, I believe this is different than the above cases since here as struct type is marked as sensitive. I think it would be enough/safe just to ignore it as long as the parameters of the struct are marked as sensitive, which we would need to handle indeed.

--- a/pkg/resource/sensitive.go
+++ b/pkg/resource/sensitive.go
@@ -36,6 +36,8 @@ const (
 	errCannotExpandWildcards          = "cannot expand wildcards"
 	errFmtCannotGetValueForFieldPath  = "cannot not get a value for fieldpath %q"
 	errFmtCannotGetStringForFieldPath = "cannot not get a string for fieldpath %q"
+	errFmtCannotGetSecretKeySelector  = "cannot get SecretKeySelector from xp resource for fieldpath %q"
+	errFmtCannotGetSecretValue        = "cannot get secret value for %v"
 )
 
 const (
@@ -44,6 +46,8 @@ const (
 	// overridden by any custom connection key configured which would break
 	// our ability to build tfstate back.
 	prefixAttribute = "attribute."
+
+	pluralSuffix = "s"
 
 	errGetAdditionalConnectionDetails = "cannot get additional connection details"
 	errFmtCannotOverrideExistingKey   = "overriding a reserved connection key (%q) is not allowed"
@@ -102,7 +106,7 @@ func GetConnectionDetails(attr map[string]interface{}, tr Terraformed, cfg *conf
 // GetSensitiveAttributes returns strings matching provided field paths in the
 // input data.
 // See the unit tests for examples.
-func GetSensitiveAttributes(from map[string]interface{}, mapping map[string]string) (map[string][]byte, error) {
+func GetSensitiveAttributes(from map[string]interface{}, mapping map[string]string) (map[string][]byte, error) { //nolint: gocyclo
 	if len(mapping) == 0 {
 		return nil, nil
 	}
@@ -124,10 +128,7 @@ func GetSensitiveAttributes(from map[string]interface{}, mapping map[string]stri
 			if v == nil {
 				continue
 			}
-			s, ok := v.(string)
-			if !ok {
-				return nil, errors.Errorf(errFmtCannotGetStringForFieldPath, fp)
-			}
+
 			// Note(turkenh): k8s secrets uses a strict regex to validate secret
 			// keys which does not allow having brackets inside. So, we need to
 			// do a conversion to be able to store as connection secret keys.
@@ -140,7 +141,21 @@ func GetSensitiveAttributes(from map[string]interface{}, mapping map[string]stri
 			if vals == nil {
 				vals = map[string][]byte{}
 			}
-			vals[fmt.Sprintf("%s%s", prefixAttribute, k)] = []byte(s)
+			switch s := v.(type) {
+			case []interface{}:
+				for i, e := range s {
+					value, ok := e.(string)
+					if !ok {
+						return nil, errors.Errorf(errFmtCannotGetStringForFieldPath, fp)
+					}
+					k = strings.TrimSuffix(k, pluralSuffix)
+					vals[fmt.Sprintf("%s%s.%v", prefixAttribute, k, i)] = []byte(value)
+				}
+			case string:
+				vals[fmt.Sprintf("%s%s", prefixAttribute, k)] = []byte(s)
+			default:
+				return nil, errors.Errorf(errFmtCannotGetStringForFieldPath, fp)
+			}
 		}
 	}
 	return vals, nil
@@ -181,20 +196,38 @@ func GetSensitiveParameters(ctx context.Context, client SecretClient, from runti
 			if v == nil {
 				continue
 			}
-			sel := &v1.SecretKeySelector{}
-			if err = pavedJSON.GetValueInto(expandedJSONPath, sel); err != nil {
-				return errors.Wrapf(err, "cannot get SecretKeySelector from xp resource for fieldpath %q", expandedJSONPath)
-			}
-			sensitive, err = client.GetSecretValue(ctx, *sel)
-			if err != nil {
-				return errors.Wrapf(err, "cannot get secret value for %v", sel)
-			}
-			expTF, err := expandedTFPath(expandedJSONPath, mapping)
-			if err != nil {
-				return err
-			}
-			if err = pavedTF.SetString(expTF, string(sensitive)); err != nil {
-				return errors.Wrapf(err, "cannot set string as terraform attribute for fieldpath %q", tfPath)
+
+			switch v.(type) {
+			case []map[string]interface{}, []interface{}:
+				sel := &[]v1.SecretKeySelector{}
+				if err = pavedJSON.GetValueInto(expandedJSONPath, sel); err != nil {
+					return errors.Wrapf(err, errFmtCannotGetSecretKeySelector, expandedJSONPath)
+				}
+				var sensitives []interface{}
+				for _, s := range *sel {
+					sensitive, err = client.GetSecretValue(ctx, s)
+					if err != nil {
+						return errors.Wrapf(err, errFmtCannotGetSecretValue, sel)
+					}
+					sensitives = append(sensitives, string(sensitive))
+				}
+				if err := setSensitiveParametersWithPaved(pavedTF, expandedJSONPath, tfPath, mapping, sensitives); err != nil {
+					return err
+				}
+			case map[string]interface{}, interface{}:
+				sel := &v1.SecretKeySelector{}
+				if err = pavedJSON.GetValueInto(expandedJSONPath, sel); err != nil {
+					return errors.Wrapf(err, errFmtCannotGetSecretKeySelector, expandedJSONPath)
+				}
+				sensitive, err = client.GetSecretValue(ctx, *sel)
+				if err != nil {
+					return errors.Wrapf(err, errFmtCannotGetSecretValue, sel)
+				}
+				if err := setSensitiveParametersWithPaved(pavedTF, expandedJSONPath, tfPath, mapping, string(sensitive)); err != nil {
+					return err
+				}
+			default:
+				return errors.Wrapf(err, errFmtCannotGetSecretKeySelector, expandedJSONPath)
 			}
 		}
 	}
@@ -326,4 +359,15 @@ func fieldPathToSecretKey(s string) (string, error) {
 	}
 
 	return strings.TrimPrefix(b.String(), "."), nil
+}
+
+func setSensitiveParametersWithPaved(pavedTF *fieldpath.Paved, expandedJSONPath, tfPath string, mapping map[string]string, sensitives interface{}) error {
+	expTF, err := expandedTFPath(expandedJSONPath, mapping)
+	if err != nil {
+		return err
+	}
+	if err = pavedTF.SetValue(expTF, sensitives); err != nil {
+		return errors.Wrapf(err, "cannot set string as terraform attribute for fieldpath %q", tfPath)
+	}
+	return nil
 }

--- a/pkg/resource/sensitive_test.go
+++ b/pkg/resource/sensitive_test.go
@@ -82,10 +82,6 @@ var (
 )
 
 func TestGetConnectionDetails(t *testing.T) {
-	testInput := map[string]interface{}{}
-	if err := json.JSParser.Unmarshal(testData, &testInput); err != nil {
-		t.Fatalf("cannot unmarshall test data: %v", err)
-	}
 	type args struct {
 		tr   Terraformed
 		cfg  *config.Resource
@@ -122,6 +118,32 @@ func TestGetConnectionDetails(t *testing.T) {
 			want: want{
 				out: map[string][]byte{
 					"attribute.top_level_secret": []byte("sensitive-data-top-level-secret"),
+				},
+			},
+		},
+		"SecretList": {
+			args: args{
+				tr: &fake.Terraformed{
+					MetadataProvider: fake.MetadataProvider{
+						ConnectionDetailsMapping: map[string]string{
+							"top_level_secrets": "status.atProvider.topLevelSecrets",
+						},
+					},
+				},
+				cfg: config.DefaultResource("terrajet_resource", nil),
+				data: map[string]interface{}{
+					"top_level_secrets": []interface{}{
+						"val1",
+						"val2",
+						"val3",
+					},
+				},
+			},
+			want: want{
+				out: map[string][]byte{
+					"attribute.top_level_secret.0": []byte("val1"),
+					"attribute.top_level_secret.1": []byte("val2"),
+					"attribute.top_level_secret.2": []byte("val3"),
 				},
 			},
 		},
@@ -413,6 +435,61 @@ func TestGetSensitiveParameters(t *testing.T) {
 				out: map[string]interface{}{
 					"some_other_key": "some_other_value",
 					"admin_password": "foo",
+				},
+			},
+		},
+		"SingleNoWildcardWithSlice": {
+			args: args{
+				clientFn: func(client *mocks.MockSecretClient) {
+					client.EXPECT().GetSecretValue(gomock.Any(), gomock.Eq(xpv1.SecretKeySelector{
+						SecretReference: xpv1.SecretReference{
+							Name:      "db-passwords",
+							Namespace: "crossplane-system",
+						},
+						Key: "admin",
+					})).Return([]byte("admin_pwd"), nil)
+					client.EXPECT().GetSecretValue(gomock.Any(), gomock.Eq(xpv1.SecretKeySelector{
+						SecretReference: xpv1.SecretReference{
+							Name:      "db-passwords",
+							Namespace: "crossplane-system",
+						},
+						Key: "system",
+					})).Return([]byte("system_pwd"), nil)
+				},
+				from: &unstructured.Unstructured{
+					Object: map[string]interface{}{
+						"spec": map[string]interface{}{
+							"forProvider": map[string]interface{}{
+								"passwordsSecretRef": []map[string]interface{}{
+									{
+										"name":      "db-passwords",
+										"namespace": "crossplane-system",
+										"key":       "admin",
+									},
+									{
+										"name":      "db-passwords",
+										"namespace": "crossplane-system",
+										"key":       "system",
+									},
+								},
+							},
+						},
+					},
+				},
+				into: map[string]interface{}{
+					"some_other_key": "some_other_value",
+				},
+				mapping: map[string]string{
+					"db_passwords": "spec.forProvider.passwordsSecretRef",
+				},
+			},
+			want: want{
+				out: map[string]interface{}{
+					"some_other_key": "some_other_value",
+					"db_passwords": []interface{}{
+						"admin_pwd",
+						"system_pwd",
+					},
 				},
 			},
 		},

--- a/pkg/types/builder.go
+++ b/pkg/types/builder.go
@@ -155,8 +155,9 @@ func (g *Builder) buildResource(res *schema.Resource, cfg *config.Resource, tfPa
 			sfx := "SecretRef"
 			cfg.Sensitive.AddFieldPath(fieldPathWithWildcard(tfPaths), "spec.forProvider."+fieldPathWithWildcard(xpPaths)+sfx)
 			// todo(turkenh): do we need to support other field types as sensitive?
-			if fieldType.String() != "string" && fieldType.String() != "*string" {
-				return nil, nil, fmt.Errorf(`got type %q for field %q, only types "string" and "*string" supported as sensitive`, fieldType.String(), fieldNameCamel)
+			if fieldType.String() != "string" && fieldType.String() != "*string" && fieldType.String() != "[]string" &&
+				fieldType.String() != "[]*string" {
+				return nil, nil, fmt.Errorf(`got type %q for field %q, only types "string", "*string", []string and []*string supported as sensitive`, fieldType.String(), fieldNameCamel)
 			}
 			// Replace a parameter field with secretKeyRef if it is sensitive.
 			// If it is an observation field, it will be dropped.
@@ -164,10 +165,15 @@ func (g *Builder) buildResource(res *schema.Resource, cfg *config.Resource, tfPa
 			fieldNameCamel += sfx
 
 			tfTag = "-"
-			fieldType = typeSecretKeySelector
+			switch fieldType.String() {
+			case "string", "*string":
+				fieldType = typeSecretKeySelector
+			case "[]string", "[]*string":
+				fieldType = types.NewSlice(typeSecretKeySelector)
+			}
 			jsonTag = name.NewFromCamel(fieldNameCamel).LowerCamelComputed
 			if sch.Optional {
-				fieldType = types.NewPointer(typeSecretKeySelector)
+				fieldType = types.NewPointer(fieldType)
 				jsonTag += ",omitempty"
 			}
 		}

